File: core/java/src/main/java/hara/lang/protocol/ISpace.java
package hara.lang.protocol;

import java.util.List;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.ispace", name = "ISpace")
public interface ISpace {
  @HaraMethod(value = "context-set", arity = 4)
  void contextSet(Object context, Object key, Object options);

  @HaraMethod(value = "context-unset", arity = 2)
  void contextUnset(Object context);

  @HaraMethod(value = "context-list", arity = 1)
  List<?> contextList();

  @HaraMethod(value = "context-get", arity = 2)
  Object contextGet(Object context);

  @HaraMethod(value = "rt-active", arity = 1)
  List<?> activeRuntimes();

  @HaraMethod(value = "rt-get", arity = 2)
  Object runtimeGet(Object context);

  @HaraMethod(value = "rt-start", arity = 2)
  Object runtimeStart(Object context);

  @HaraMethod(value = "rt-started?", arity = 2)
  boolean runtimeStarted(Object context);

  @HaraMethod(value = "rt-stopped?", arity = 2)
  boolean runtimeStopped(Object context);

  @HaraMethod(value = "rt-stop", arity = 2)
  void runtimeStop(Object context);
}

File: core/rust/src/lang/protocol/ispace.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.ispace", name = "ISpace")]
pub trait ISpace<C, K, O> {
    type Runtime;

    #[hara_method(value = "context-set", arity = 4)]
    fn context_set(&mut self, context: C, key: K, options: O);
    #[hara_method(value = "context-unset", arity = 2)]
    fn context_unset(&mut self, context: &C);
    #[hara_method(value = "context-list", arity = 1)]
    fn context_list(&self) -> Vec<C>;
    #[hara_method(value = "context-get", arity = 2)]
    fn context_get(&self, context: &C) -> Option<O>;
    #[hara_method(value = "rt-active", arity = 1)]
    fn runtime_active(&self) -> Vec<Self::Runtime>;
    #[hara_method(value = "rt-get", arity = 2)]
    fn runtime_get(&self, context: &C) -> Option<Self::Runtime>;
    #[hara_method(value = "rt-start", arity = 2)]
    fn runtime_start(&mut self, context: C) -> Self::Runtime;
    #[hara_method(value = "rt-started?", arity = 2)]
    fn runtime_started(&self, context: &C) -> bool;
    #[hara_method(value = "rt-stopped?", arity = 2)]
    fn runtime_stopped(&self, context: &C) -> bool;
    #[hara_method(value = "rt-stop", arity = 2)]
    fn runtime_stop(&mut self, context: &C);
}
